File: opennotes-discord/src/cache/adapters/redis.ts
import Redis, { type RedisOptions } from 'ioredis';
import { logger } from '../../logger.js';
import { sanitizeConnectionUrl } from '../../utils/url-sanitizer.js';
import { safeJSONParse, safeJSONStringify } from '../../utils/safe-json.js';
import type { CacheConfig, CacheInterface, CacheMetrics } from '../interfaces.js';

export interface RedisCacheConfig extends CacheConfig {
  url?: string;
  host?: string;
  port?: number;
  password?: string | undefined;
  db?: number;
  maxRetries?: number;
  connectTimeout?: number;
}

/**
 * Redis cache implementation with connection pooling and retry logic
 *
 * Features:
 * - Automatic connection pooling
 * - Exponential backoff retry strategy
 * - Graceful error handling and reconnection
 * - Pub/Sub support for cache invalidation
 * - Circuit breaker pattern
 * - Metrics collection
 *
 * Suitable for:
 * - Production deployments
 * - Multi-instance/distributed systems
 * - High-availability requirements
 */
interface ResolvedRedisCacheConfig extends Required<Omit<RedisCacheConfig, 'password'>> {
  password: string | undefined;
}

export class RedisCacheAdapter implements CacheInterface {
  private client: Redis;
  private config: ResolvedRedisCacheConfig;
  private metrics: CacheMetrics;
  private isConnected: boolean;
  private reconnecting: boolean;
  private subscribers: Map<string, Redis> = new Map();

  constructor(config: RedisCacheConfig = {}) {
    this.isConnected = false;
    this.reconnecting = false;

    this.config = {
      defaultTtl: config.defaultTtl ?? 300,
      keyPrefix: config.keyPrefix ?? '',
      maxSize: config.maxSize ?? 10000,
      evictionPolicy: config.evictionPolicy ?? 'lru',
      url: config.url ?? process.env.REDIS_URL ?? 'redis://localhost:6379',
      host: config.host ?? 'localhost',
      port: config.port ?? 6379,
      password: config.password ?? undefined,
      db: config.db ?? 0,
      maxRetries: config.maxRetries ?? 10,
      connectTimeout: config.connectTimeout ?? 10000,
    };

    this.metrics = {
      hits: 0,
      misses: 0,
      sets: 0,
      deletes: 0,
      evictions: 0,
      size: 0,
    };

    const redisOptions: RedisOptions = {
      host: this.config.host,
      port: this.config.port,
      password: this.config.password,
      db: this.config.db,
      connectTimeout: this.config.connectTimeout,
      maxRetriesPerRequest: 3,
      enableOfflineQueue: true,
      lazyConnect: true,
      retryStrategy: (times: number) => {
        if (times > this.config.maxRetries) {
          logger.error('Redis max retries exceeded', { times });
          return null;
        }
        const delay = Math.min(times * 50, 2000);
        logger.debug('Redis retry', { times, delay });
        return delay;
      },
      reconnectOnError: (err: Error) => {
        const targetErrors = ['READONLY', 'ECONNREFUSED', 'ETIMEDOUT'];
        if (targetErrors.some((target) => err.message.includes(target))) {
          logger.warn('Redis reconnecting on error', { error: err.message });
          return 2;
        }
        return false;
      },
    };

    if (this.config.url) {
      if (this.config.url.startsWith('rediss://')) {
        redisOptions.tls = {
          rejectUnauthorized: false,
        };
        logger.debug('Redis TLS enabled for rediss:// connection');
      }
      this.client = new Redis(this.config.url, redisOptions);
    } else {
      this.client = new Redis(redisOptions);
    }

    this.client.on('connect', () => {
      this.isConnected = true;
      this.reconnecting = false;
      const connectionUrl = this.config.url
        ? sanitizeConnectionUrl(this.config.url)
        : `redis://${this.config.host}:${this.config.port}`;
      logger.info('Redis connected', {
        url: connectionUrl,
        db: this.config.db,
      });
    });

    this.client.on('ready', () => {
      logger.info('Redis ready');
    });

    this.client.on('error', (err: unknown) => {
      const error = err instanceof Error ? err.message : String(err);
      logger.error('Redis error', { error });
    });

    this.client.on('close', () => {
      this.isConnected = false;
      logger.warn('Redis connection closed');
    });

    this.client.on('reconnecting', () => {
      this.reconnecting = true;
      logger.info('Redis reconnecting');
    });

    this.client.on('end', () => {
      this.isConnected = false;
      logger.warn('Redis connection ended');
    });
  }

  start(): void {
    if (!this.isConnected && !this.reconnecting) {
      void this.client.connect().catch((err: unknown) => {
        const error = err instanceof Error ? err.message : String(err);
        logger.error('Redis connection failed', { error });
      });
    }
  }

  stop(): void {
    for (const [channel, subscriber] of this.subscribers.entries()) {
      try {
        void subscriber.unsubscribe(channel);
        void subscriber.disconnect();
      } catch (err: unknown) {
        const error = err instanceof Error ? err.message : String(err);
        logger.error('Error cleaning up subscriber', { channel, error });
      }
    }
    this.subscribers.clear();
    logger.info('Redis subscribers cleaned up', { count: this.subscribers.size });

    this.client.removeAllListeners();

    if (this.isConnected || this.reconnecting) {
      this.client.disconnect();
      logger.info('Redis disconnected');
    }
  }

  async get<T>(key: string): Promise<T | null> {
    try {
      const fullKey = this.buildKey(key);
      const value = await this.client.get(fullKey);

      if (value === null) {
        this.metrics.misses++;
        return null;
      }

      this.metrics.hits++;
      return safeJSONParse<T>(value);
    } catch (error) {
      logger.error('Redis get error', { key, error });
      this.metrics.misses++;
      return null;
    }
  }

  async set(key: string, value: unknown, ttl?: number): Promise<boolean> {
    try {
      const fullKey = this.buildKey(key);
      const effectiveTtl = ttl ?? this.config.defaultTtl;
      const serialized = safeJSONStringify(value);

      if (effectiveTtl > 0) {
        await this.client.setex(fullKey, effectiveTtl, serialized);
      } else {
        await this.client.set(fullKey, serialized);
      }

      this.metrics.sets++;
      return true;
    } catch (error) {
      logger.error('Redis set error', { key, error });
      return false;
    }
  }

  async delete(key: string): Promise<boolean> {
    try {
      const fullKey = this.buildKey(key);
      const result = await this.client.del(fullKey);
      const existed = result > 0;

      if (existed) {
        this.metrics.deletes++;
      }

      return existed;
    } catch (error) {
      logger.error('Redis delete error', { key, error });
      return false;
    }
  }

  async exists(key: string): Promise<boolean> {
    try {
      const fullKey = this.buildKey(key);
      const result = await this.client.exists(fullKey);
      return result === 1;
    } catch (error) {
      logger.error('Redis exists error', { key, error });
      return false;
    }
  }

  async expire(key: string, ttl: number): Promise<boolean> {
    try {
      const fullKey = this.buildKey(key);
      const result = await this.client.expire(fullKey, ttl);
      return result === 1;
    } catch (error) {
      logger.error('Redis expire error', { key, error });
      return false;
    }
  }

  async mget(keys: string[]): Promise<(string | null)[]> {
    if (keys.length === 0) {
      return [];
    }

    try {
      const fullKeys = keys.map((k) => this.buildKey(k));
      const values = await this.client.mget(...fullKeys);

      return values.map((v: string | null) => {
        if (v === null) {
          this.metrics.misses++;
          return null;
        }
        this.metrics.hits++;
        try {
          return safeJSONParse(v);
        } catch {
          return null;
        }
      });
    } catch (error) {
      logger.error('Redis mget error', { keys, error });
      return keys.map(() => null);
    }
  }

  async mset(items: Map<string, unknown>, ttl?: number): Promise<boolean> {
    if (items.size === 0) {
      return true;
    }

    try {
      const pipeline = this.client.pipeline();
      const effectiveTtl = ttl ?? this.config.defaultTtl;

      for (const [key, value] of items.entries()) {
        const fullKey = this.buildKey(key);
        const serialized = safeJSONStringify(value);

        if (effectiveTtl > 0) {
          pipeline.setex(fullKey, effectiveTtl, serialized);
        } else {
          pipeline.set(fullKey, serialized);
        }
      }

      await pipeline.exec();
      this.metrics.sets += items.size;
      return true;
    } catch (error) {
      logger.error('Redis mset error', { count: items.size, error });
      return false;
    }
  }

  async clear(pattern?: string): Promise<number> {
    try {
      let cursor = '0';
      let count = 0;
      const matchPattern = pattern
        ? this.buildKey(pattern)
        : this.buildKey('*');

      do {
        const [nextCursor, keys] = await this.client.scan(
          cursor,
          'MATCH',
          matchPattern,
          'COUNT',
          100
        );
        cursor = nextCursor;

        if (keys.length > 0) {
          const deleted = await this.client.del(...keys);
          count += deleted;
        }
      } while (cursor !== '0');

      return count;
    } catch (error) {
      logger.error('Redis clear error', { pattern, error });
      return 0;
    }
  }

  async ping(): Promise<boolean> {
    try {
      const result = await this.client.ping();
      return result === 'PONG';
    } catch (error) {
      logger.error('Redis ping error', { error });
      return false;
    }
  }

  getMetrics(): CacheMetrics {
    return { ...this.metrics };
  }

  private buildKey(key: string): string {
    return this.config.keyPrefix ? `${this.config.keyPrefix}:${key}` : key;
  }

  async subscribe(channel: string, handler: (message: string) => void): Promise<void> {
    let subscriber = this.subscribers.get(channel);

    if (!subscriber) {
      subscriber = this.client.duplicate();
      try {
        await subscriber.subscribe(channel);
        this.subscribers.set(channel, subscriber);
        logger.info('Redis subscribed', { channel, totalSubscribers: this.subscribers.size });
      } catch (err: unknown) {
        const error = err instanceof Error ? err.message : String(err);
        logger.error('Redis subscribe error', { channel, error });
        throw err;
      }
    }

    subscriber.on('message', (ch: string, message: string) => {
      if (ch === channel) {
        handler(message);
      }
    });
  }

  async unsubscribe(channel: string): Promise<void> {
    const subscriber = this.subscribers.get(channel);
    if (subscriber) {
      try {
        await subscriber.unsubscribe(channel);
        subscriber.disconnect();
        this.subscribers.delete(channel);
        logger.info('Redis unsubscribed', { channel, remainingSubscribers: this.subscribers.size });
      } catch (err: unknown) {
        const error = err instanceof Error ? err.message : String(err);
        logger.error('Redis unsubscribe error', { channel, error });
      }
    }
  }

  async publish(channel: string, message: string): Promise<number> {
    try {
      return await this.client.publish(channel, message);
    } catch (error) {
      logger.error('Redis publish error', { channel, error });
      return 0;
    }
  }
}

File: opennotes-server/src/cache/adapters/redis.py
"""
Redis cache implementation with connection pooling and retry logic.

Features:
- Async operations with connection pooling
- Automatic retry with exponential backoff
- Connection health monitoring
- Pub/Sub support for cache invalidation
- Circuit breaker pattern
- Metrics collection

Suitable for:
- Production deployments
- Multi-instance/distributed systems
- High-availability requirements
"""

import asyncio
import hashlib
import hmac
import json
import logging
import time
from collections.abc import Callable
from typing import Any, Generic, TypeVar, cast

import orjson
from pydantic import BaseModel, Field, ValidationError
from redis.asyncio import Redis
from redis.asyncio.connection import ConnectionPool
from redis.asyncio.retry import Retry
from redis.backoff import ExponentialBackoff
from redis.exceptions import (
    ConnectionError as RedisConnectionError,
)
from redis.exceptions import (
    RedisError,
)
from redis.exceptions import (
    TimeoutError as RedisTimeoutError,
)

from src.config import settings

from ..interfaces import CacheConfig, CacheInterface, CacheMetrics
from ..redis_client import get_redis_connection_kwargs

logger = logging.getLogger(__name__)
T = TypeVar("T")

MAX_MESSAGE_AGE_SECONDS = 300


class PubSubMessage(BaseModel):
    """Validated pub/sub message structure."""

    type: str = Field(..., min_length=1, max_length=100)
    payload: dict[str, Any] | str
    timestamp: int = Field(..., ge=0)
    signature: str | None = Field(default=None, min_length=64, max_length=128)


class RedisCacheAdapter(CacheInterface[T], Generic[T]):
    """
    Redis cache adapter with connection pooling and retry logic.

    Type Parameters:
        T: The type of values stored in the cache

    Example:
        # Cache with JSON-serializable data
        cache: RedisCacheAdapter[dict[str, Any]] = RedisCacheAdapter(url="redis://localhost")
        await cache.start()
        await cache.set("session:123", {"user_id": 456, "expires": 1234567890})
        session: dict[str, Any] | None = await cache.get("session:123")

        # Cache with Pydantic models (using schema validation)
        cache: RedisCacheAdapter[UserModel] = RedisCacheAdapter(url="redis://localhost")
        await cache.start()
        await cache.set("user:123", user_model)
        user: UserModel | None = await cache.get("user:123", schema=UserModel)
    """

    def __init__(
        self,
        config: CacheConfig | None = None,
        url: str | None = None,
        host: str = "localhost",
        port: int = 6379,
        password: str | None = None,
        db: int = 0,
        max_connections: int = 10,
        max_retries: int = 3,
        socket_timeout: float = 5.0,
        socket_connect_timeout: float = 5.0,
    ):
        self.config = config or CacheConfig()
        self.metrics = CacheMetrics()

        # Connection parameters - use settings for consistency
        self.url = url or settings.REDIS_URL
        self.host = host
        self.port = port
        self.password = password
        self.db = db
        self.max_connections = max_connections
        self.max_retries = max_retries
        self.socket_timeout = socket_timeout
        self.socket_connect_timeout = socket_connect_timeout

        # Connection objects
        self.pool: ConnectionPool | None = None
        self.client: Redis | None = None
        self.is_connected = False

        # Track subscription tasks to prevent leaks
        self.subscription_tasks: list[asyncio.Task[None]] = []

    async def start(self) -> None:
        """Initialize Redis connection pool."""
        if self.is_connected:
            return

        try:
            # Ensure url is not None
            if self.url is None:
                raise ValueError("Redis URL cannot be None")

            # Create connection pool with retry strategy
            retry = Retry(ExponentialBackoff(), self.max_retries)

            # Get base connection kwargs from shared factory (handles SSL, TLS validation)
            pool_kwargs = get_redis_connection_kwargs(
                self.url,
                decode_responses=False,  # We handle JSON serialization
                max_connections=self.max_connections,
                socket_timeout=self.socket_timeout,
                socket_connect_timeout=self.socket_connect_timeout,
                retry=retry,
            )

            self.pool = ConnectionPool.from_url(self.url, **pool_kwargs)

            self.client = Redis(connection_pool=self.pool)

            # Test connection
            await self.client.ping()
            self.is_connected = True

        except (RedisConnectionError, RedisTimeoutError, RedisError) as e:
            self.is_connected = False
            # Clean up resources on failure to prevent leaks
            if self.client:
                try:
                    await self.client.aclose()
                except RedisError:
                    pass
                self.client = None
            if self.pool:
                try:
                    await self.pool.aclose()
                except RedisError:
                    pass
                self.pool = None
            raise RedisConnectionError(f"Failed to connect to Redis: {e}") from e

    async def stop(self) -> None:
        """Close Redis connection and cleanup resources."""
        if not self.is_connected:
            return

        try:
            # Cancel all subscription tasks first
            for task in self.subscription_tasks:
                if not task.done():
                    task.cancel()

            # Wait for all tasks to complete or be cancelled
            if self.subscription_tasks:
                await asyncio.gather(*self.subscription_tasks, return_exceptions=True)
                self.subscription_tasks.clear()

            if self.client:
                await self.client.aclose()
                self.client = None

            if self.pool:
                await self.pool.aclose()
                self.pool = None

            self.is_connected = False

        except RedisError:
            pass

    async def _ensure_connected(self) -> bool:
        """Ensure Redis connection is alive and reconnect if needed."""
        if not self.is_connected or not self.client:
            try:
                await self.start()
                return True
            except (RedisConnectionError, RedisTimeoutError, RedisError):
                return False

        try:
            await self.client.ping()
            return True
        except (RedisConnectionError, RedisTimeoutError, RedisError):
            self.is_connected = False
            try:
                await self.start()
                return True
            except (RedisConnectionError, RedisTimeoutError, RedisError):
                return False

    async def get(  # noqa: PLR0911
        self,
        key: str,
        default: T | None = None,
        value_type: type[T] | None = None,
        schema: type[BaseModel] | None = None,
    ) -> T | None:
        """Get a value from the cache with optional type and schema validation."""
        if not await self._ensure_connected():
            return default

        assert self.client is not None  # Type narrowing after connection check

        try:
            full_key = self._build_key(key)
            value = await self.client.get(full_key)

            if value is None:
                self.metrics.misses += 1
                return default

            deserialized = orjson.loads(value)

            if schema is not None:
                try:
                    validated = schema.model_validate(deserialized)
                    self.metrics.hits += 1
                    return cast(T, validated)
                except ValidationError as e:
                    logger.error(
                        f"Cache validation failed for key {key}: {e.error_count()} errors",
                        extra={"key": key, "errors": e.errors()},
                    )
                    self.metrics.misses += 1
                    return default

            if value_type is not None and not isinstance(deserialized, value_type):
                logger.warning(
                    f"Type mismatch for key {key}: expected {value_type.__name__}, "
                    f"got {type(deserialized).__name__}"
                )
                self.metrics.misses += 1
                return default

            self.metrics.hits += 1
            return cast(T, deserialized)

        except orjson.JSONDecodeError as e:
            logger.warning(
                f"Failed to deserialize cached value for key '{key}': {e}",
                extra={"key": key, "error_type": "json_decode"},
            )
            self.metrics.misses += 1
            return default
        except RedisError as e:
            logger.error(
                f"Redis error getting key '{key}': {e}",
                extra={"key": key, "error_type": "redis"},
            )
            self.metrics.misses += 1
            return default

    async def set(self, key: str, value: T, ttl: int | None = None) -> bool:
        """Set a value in the cache."""
        if not await self._ensure_connected():
            return False

        assert self.client is not None  # Type narrowing after connection check

        try:
            full_key = self._build_key(key)
            effective_ttl = ttl if ttl is not None else self.config.default_ttl

            if effective_ttl < 0:
                logger.warning(
                    f"Invalid TTL value {effective_ttl} for key '{key}', using 0 (no expiration)"
                )
                effective_ttl = 0

            serialized = orjson.dumps(value)

            if effective_ttl > 0:
                await self.client.setex(full_key, effective_ttl, serialized)
            else:
                await self.client.set(full_key, serialized)

            self.metrics.sets += 1
            return True

        except TypeError as e:
            logger.error(
                f"Failed to serialize value for key '{key}': {e}",
                extra={"key": key, "error_type": "json_encode", "value_type": type(value).__name__},
            )
            return False
        except RedisError as e:
            logger.error(
                f"Redis error setting key '{key}': {e}",
                extra={"key": key, "error_type": "redis"},
            )
            return False

    async def delete(self, key: str) -> bool:
        """Delete a value from the cache."""
        if not await self._ensure_connected():
            return False

        assert self.client is not None  # Type narrowing after connection check

        try:
            full_key = self._build_key(key)
            result = await self.client.delete(full_key)
            existed = bool(result > 0)

            if existed:
                self.metrics.deletes += 1

            return existed

        except RedisError:
            return False

    async def exists(self, key: str) -> bool:
        """Check if a key exists in the cache."""
        if not await self._ensure_connected():
            return False

        assert self.client is not None  # Type narrowing after connection check

        try:
            full_key = self._build_key(key)
            result = await self.client.exists(full_key)
            return bool(result == 1)

        except RedisError:
            return False

    async def expire(self, key: str, ttl: int) -> bool:
        """Set expiration time for a key."""
        if not await self._ensure_connected():
            return False

        assert self.client is not None  # Type narrowing after connection check

        try:
            full_key = self._build_key(key)
            result = await self.client.expire(full_key, ttl)
            return bool(result == 1)

        except RedisError:
            return False

    async def mget(self, keys: list[str], schema: type[BaseModel] | None = None) -> list[T | None]:
        """Get multiple values from the cache with optional schema validation."""
        if not keys:
            return []
        if not await self._ensure_connected():
            return [None] * len(keys)

        assert self.client is not None  # Type narrowing after connection check

        try:
            full_keys = [self._build_key(k) for k in keys]
            values = await self.client.mget(full_keys)

            result: list[T | None] = []
            for idx, value in enumerate(values):
                if value is None:
                    self.metrics.misses += 1
                    result.append(None)
                else:
                    try:
                        deserialized = orjson.loads(value)

                        if schema is not None:
                            try:
                                validated = schema.model_validate(deserialized)
                                self.metrics.hits += 1
                                result.append(cast(T, validated))
                            except ValidationError as e:
                                logger.error(
                                    f"Cache validation failed for key {keys[idx]}: {e.error_count()} errors"
                                )
                                self.metrics.misses += 1
                                result.append(None)
                        else:
                            self.metrics.hits += 1
                            result.append(cast(T, deserialized))
                    except orjson.JSONDecodeError:
                        self.metrics.misses += 1
                        result.append(None)

            return result

        except RedisError:
            self.metrics.misses += len(keys)
            return [None] * len(keys)

    async def mset(self, items: dict[str, T], ttl: int | None = None) -> bool:
        """Set multiple values in the cache."""
        if not items:
            return False
        if not await self._ensure_connected():
            return False

        assert self.client is not None  # Type narrowing after connection check

        try:
            effective_ttl = ttl if ttl is not None else self.config.default_ttl

            # Use pipeline for atomic operations
            async with self.client.pipeline(transaction=False) as pipe:
                for key, value in items.items():
                    full_key = self._build_key(key)
                    serialized = orjson.dumps(value)

                    if effective_ttl > 0:
                        pipe.setex(full_key, effective_ttl, serialized)
                    else:
                        pipe.set(full_key, serialized)

                await pipe.execute()

            self.metrics.sets += len(items)
            return True

        except (RedisError, orjson.JSONDecodeError):
            return False

    async def clear(self, pattern: str | None = None) -> int:
        """Clear cache entries matching a pattern."""
        if not await self._ensure_connected():
            return 0

        assert self.client is not None  # Type narrowing after connection check

        try:
            match_pattern = self._build_key(pattern) if pattern else self._build_key("*")

            count = 0
            cursor = 0

            while True:
                cursor, keys = await self.client.scan(cursor, match=match_pattern, count=100)

                if keys:
                    deleted = await self.client.delete(*keys)
                    count += deleted

                if cursor == 0:
                    break

            return count

        except RedisError:
            return 0

    async def ping(self) -> bool:
        """Check if the cache is healthy."""
        return await self._ensure_connected()

    def get_metrics(self) -> CacheMetrics:
        """Get cache metrics."""
        return self.metrics

    def _build_key(self, key: str) -> str:
        """Build full key with prefix."""
        return f"{self.config.key_prefix}:{key}" if self.config.key_prefix else key

    async def subscribe(
        self,
        channel: str,
        handler: Callable[[PubSubMessage], None],
        hmac_secret: str | None = None,
    ) -> None:
        """Subscribe to a Redis pub/sub channel with message validation."""
        if not await self._ensure_connected():
            return

        assert self.client is not None  # Type narrowing after connection check

        try:
            pubsub = self.client.pubsub()
            await pubsub.subscribe(channel)

            # Listen for messages in background
            async def listen() -> None:
                try:
                    async for message in pubsub.listen():
                        if message["type"] == "message":
                            raw_data = message["data"].decode("utf-8")

                            if not self._validate_message(raw_data, hmac_secret):
                                logger.warning(
                                    f"Invalid pub/sub message received on channel {channel}"
                                )
                                continue

                            try:
                                validated_message = PubSubMessage.model_validate_json(raw_data)
                                handler(validated_message)
                            except ValidationError as e:
                                logger.error(
                                    f"Failed to parse/validate message on channel {channel}: {e}"
                                )
                                continue

                except asyncio.CancelledError:
                    # Clean shutdown of subscription
                    await pubsub.unsubscribe(channel)
                    await pubsub.aclose()  # type: ignore[no-untyped-call]
                    raise

            # Store task reference to prevent garbage collection and enable cancellation
            task = asyncio.create_task(listen())
            self.subscription_tasks.append(task)

        except RedisError:
            pass

    def _validate_message(self, raw_data: str, hmac_secret: str | None = None) -> bool:
        """Validate message structure, timestamp, and optional HMAC signature."""
        try:
            message = PubSubMessage.model_validate_json(raw_data)

            current_time = int(time.time())
            message_age = abs(current_time - message.timestamp)

            if message_age > MAX_MESSAGE_AGE_SECONDS:
                logger.warning(
                    f"Message timestamp too old: age={message_age}s, max={MAX_MESSAGE_AGE_SECONDS}s"
                )
                return False

            if hmac_secret and message.signature:
                payload_str = json.dumps(message.payload, sort_keys=True)
                message_to_sign = f"{message.type}:{payload_str}:{message.timestamp}"
                expected_signature = hmac.new(
                    hmac_secret.encode(),
                    message_to_sign.encode(),
                    hashlib.sha256,
                ).hexdigest()

                if not hmac.compare_digest(expected_signature, message.signature):
                    logger.warning("Message HMAC signature verification failed")
                    return False

            return True

        except (orjson.JSONDecodeError, ValidationError) as e:
            logger.error(f"Message validation failed: {e}")
            return False

    async def publish(self, channel: str, message: str) -> int:
        """Publish a message to a Redis pub/sub channel."""
        if not await self._ensure_connected():
            return 0

        assert self.client is not None  # Type narrowing after connection check

        try:
            result = await self.client.publish(channel, message.encode("utf-8"))
            return int(result)

        except RedisError:
            return 0
